File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.py
from typing import List, Optional

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def nextLargerNodes(self, head: Optional[ListNode]) -> List[int]:
        # Initialize an empty stack and an empty result list
        stack, res = [], []
        
        # Traverse the linked list
        while head:
            # Append the node value to the result list
            res.append(0)
            
            # While the stack is not empty and the top element is smaller than the current value
            while stack and stack[-1][1] < head.val:
                # Update the corresponding index in the result list
                idx, _ = stack.pop()
                res[idx] = head.val
                
            # Append the index and value to the stack
            stack.append((len(res) - 1, head.val))
            
            # Move to the next node
            head = head.next
            
        return res

File: 1019-next-greater-node-in-linked-list/1019-next-greater-node-in-linked-list.c
// Definition for singly-linked list.
// struct ListNode {
//     int val;
//     struct ListNode *next;
// };

int* nextLargerNodes(struct ListNode* head, int* returnSize){
    // Initialize stack and result array
    int stack[10000][2], top = -1, i = 0;
    int *res = (int*)malloc(10000 * sizeof(int));
    
    while (head) {
        res[i] = 0;
        while (top >= 0 && stack[top][1] < head->val) {
            res[stack[top][0]] = head->val;
            top--;
        }
        stack[++top][0] = i;
        stack[top][1] = head->val;
        head = head->next;
        i++;
    }
    
    *returnSize = i;
    return res;
}
